Source/WebKit/WebProcess/WebPage/gtk/AcceleratedSurfaceDMABuf.cpp

```diff
@@ -0,0 +1,134 @@
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "AcceleratedSurfaceDMABuf.h"
+
+#if USE(EGL)
+#include "AcceleratedBackingStoreDMABufMessages.h"
+#include "WebPage.h"
+#include <WebCore/GLContextEGL.h>
+#include <WebCore/PlatformDisplay.h>
+#include <epoxy/egl.h>
+
+namespace WebKit {
+
+std::unique_ptr<AcceleratedSurfaceDMABuf> AcceleratedSurfaceDMABuf::create(WebPage& webPage, Client& client)
+{
+    return std::unique_ptr<AcceleratedSurfaceDMABuf>(new AcceleratedSurfaceDMABuf(webPage, client));
+}
+
+AcceleratedSurfaceDMABuf::AcceleratedSurfaceDMABuf(WebPage& webPage, Client& client)
+    : AcceleratedSurface(webPage, client)
+{
+}
+
+AcceleratedSurfaceDMABuf::~AcceleratedSurfaceDMABuf()
+{
+}
+
+void AcceleratedSurfaceDMABuf::didCreateGLContext()
+{
+    glGenTextures(1, &m_texture);
+    glBindTexture(GL_TEXTURE_2D, m_texture);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+
+    glGenFramebuffers(1, &m_fbo);
+    glBindFramebuffer(GL_FRAMEBUFFER, m_fbo);
+
+    clientResize(m_size);
+}
+
+void AcceleratedSurfaceDMABuf::willDestroyGLContext()
+{
+    if (m_texture) {
+        glDeleteTextures(1, &m_texture);
+        m_texture = 0;
+    }
+
+    if (m_fbo) {
+        glDeleteFramebuffers(1, &m_fbo);
+        m_fbo = 0;
+    }
+}
+
+uint64_t AcceleratedSurfaceDMABuf::surfaceID() const
+{
+    return m_webPage.identifier().toUInt64();
+}
+
+void AcceleratedSurfaceDMABuf::clientResize(const WebCore::IntSize& size)
+{
+    if (size.isEmpty())
+        return;
+
+    glBindTexture(GL_TEXTURE_2D, m_texture);
+    glBindFramebuffer(GL_FRAMEBUFFER, m_fbo);
+    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, size.width(), size.height(), 0, GL_RGBA, GL_UNSIGNED_BYTE, 0);
+    glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, m_texture, 0);
+    glFlush();
+
+    exportDMABuf(size);
+}
+
+void AcceleratedSurfaceDMABuf::exportDMABuf(const WebCore::IntSize& size)
+{
+    auto* context = GLContext::current();
+    ASSERT(context && context->isEGLContext());
+    auto& eglContext = downcast<GLContextEGL>(*context);
+    auto image = eglContext.createImage(EGL_GL_TEXTURE_2D, (EGLClientBuffer)(uint64_t)m_texture, { });
+    auto eglDisplay = WebCore::PlatformDisplay::sharedDisplayForCompositing().eglDisplay();
+    int fd, fourcc;
+    EGLint offset, stride;
+    eglExportDMABUFImageQueryMESA(eglDisplay, image, &fourcc, nullptr, nullptr);
+    eglExportDMABUFImageMESA(eglDisplay, image, &fd, &stride, &offset);
```

Comment:
The extension itself is not widely supported. There's also no control over the parameters of the underlying dmabuf, meaning the internal implementation can select some combination of the format and modifier values that might cause problems during importing on the other side, because e.g. modifier information is not available or the GTK-controlled context ends up in a configuration where the import doesn't work.

Current approach also implements rendering into a single dmabuf, which is then sampled from in the UIProcess. This can cause problems if WebProcess renders into the dmabuf as it's being sampled in the UIProcess. Simple double-buffering with commit-release mechanics would work around that. Better yet, a GBMBufferSwapchain can be used similarly to how it's used in GraphicsContextGLGBM.

This approach was already implemented in WPEBackend-fdo, as the dmabuf-pooling variant. The rendering backend implementation is here:
https://github.com/Igalia/WPEBackend-fdo/blob/master/src/egl-client-dmabuf-pool.cpp